### ISSUE TITLE
Serialization fails when @transient val is passed to constructor

### ISSUE THREAD

**Issue by ekrich** [lampepfl/dotty#9881]
## Minimized code

```Scala
import java.io._

class Config(s: String)

class ConfigException(@transient val config: Config = null)
    extends java.io.Serializable {}

object Main {

  def main(args: Array[String]): Unit = {
    val e =
      new ConfigException(new Config("not serializable"))
    val byteStream = new ByteArrayOutputStream()
    val objectStream = new ObjectOutputStream(byteStream)
    objectStream.writeObject(e)
    objectStream.close()
    val bytes = byteStream.toByteArray()
    val inStream = new ByteArrayInputStream(bytes)
    val inObjectStream = new ObjectInputStream(inStream)
    val copy = inObjectStream.readObject()
    inObjectStream.close()
  }
}

```

## Output

```scala
sbt:dotty-simple> run
[info] Compiling 1 Scala source to /Users/eric/workspace/dotty-test/target/scala-0.27/classes ...
[info] running Main 
[error] (run-main-1a) java.io.NotSerializableException: Config
[error] java.io.NotSerializableException: Config
[error] 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
[error] 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[error] 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[error] 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[error] 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[error] 	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
[error] 	at Main$.main(Main.scala:15)
[error] 	at Main.main(Main.scala)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
[error] Total time: 0 s, completed Sep 25, 2020 11:33:00 AM
sbt:dotty-simple> 
```

## Expectation

In Scala 2, serialization avoids the `@transient val` so I am thinking it should avoid in Scala 3 too.
Note: If the `@transient val or var` is in the body of the class, Scala 3 does not try to serialize.

**Comment by ekrich** [comment]
Here is the info about the class. I don't see any flags for `@transient`, `ACC_TRANSIENT`? so I think it may be just omitted.
```
$ javap -verbose Config.class
Classfile /<removed>/dotty-test/target/scala-0.27/classes/Config.class
  Last modified Sep 25, 2020; size 345 bytes
  MD5 checksum 685a6ce3fea538183b6870df70e61056
  Compiled from "Main.scala"
public class Config
  minor version: 0
  major version: 52
  flags: ACC_PUBLIC, ACC_SUPER
Constant pool:
   #1 = Utf8               Config
   #2 = Class              #1             // Config
   #3 = Utf8               java/lang/Object
   #4 = Class              #3             // java/lang/Object
   #5 = Utf8               Main.scala
   #6 = Utf8               <init>
   #7 = Utf8               (Ljava/lang/String;)V
   #8 = Utf8               ()V
   #9 = NameAndType        #6:#8          // "<init>":()V
  #10 = Methodref          #4.#9          // java/lang/Object."<init>":()V
  #11 = Utf8               this
  #12 = Utf8               LConfig;
  #13 = Utf8               s
  #14 = Utf8               Ljava/lang/String;
  #15 = Utf8               Code
  #16 = Utf8               LineNumberTable
  #17 = Utf8               LocalVariableTable
  #18 = Utf8               SourceFile
  #19 = Utf8               TASTY
  #20 = Utf8               Scala
{
  public Config(java.lang.String);
    descriptor: (Ljava/lang/String;)V
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=2, args_size=2
         0: aload_0
         1: invokespecial #10                 // Method java/lang/Object."<init>":()V
         4: return
      LineNumberTable:
        line 3: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       5     0  this   LConfig;
            0       5     1     s   Ljava/lang/String;
}
SourceFile: "Main.scala"
Error: unknown attribute
  TASTY: length = 0x10
   00 2D D3 B7 5A AA A4 00 00 97 31 7E 19 E7 18 00

Error: unknown attribute
  Scala: length = 0x0
```

**Comment by ekrich** [comment]
Ok, I have confirmed that if the `@transient` is in the body we get the following:
```
private final transient Config config;
    descriptor: LConfig;
    flags: ACC_PRIVATE, ACC_FINAL, ACC_TRANSIENT
```
If the `@transient` is in the constructor we get the following:
```
private final Config config;
    descriptor: LConfig;
    flags: ACC_PRIVATE, ACC_FINAL
```

**Comment by smarter** [comment]
`@transient` is defined with the meta-annotation `@field` which lets the compiler know that it needs to be put on the field: https://github.com/scala/scala/blob/8c86b7d7136839538cca0ff8fca50f59437564c0/src/library/scala/transient.scala#L17-L18, but we have almost no support for meta-annotations currently (see https://github.com/scala/scala/blob/2.13.x/src/library/scala/annotation/meta/package.scala for how they're supposed to work) so this isn't trivial to fix. Here's a sketch of what should be done:
- in https://github.com/lampepfl/dotty/blob/7f46c251362a86373184e64216b094745f09a704/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L436 we do `keepAnnotations = true` for the constructor parameters but this is incorrect if meta-annotations exist for the current annotation unless the meta-annotaton `@param` is part of these meta-annotations.
- in https://github.com/lampepfl/dotty/blob/7f46c251362a86373184e64216b094745f09a704/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L787-L788 all annotations are dropped from the accessors, but we need to keep the ones which have `@field`, `@getter` or `@setter` as a meta-annotation
- in https://github.com/lampepfl/dotty/blob/7f46c251362a86373184e64216b094745f09a704/compiler/src/dotty/tools/dotc/transform/Memoize.scala accessors are split into field/getter/setter, this is where we need to filter which annotations go where depending on their meta-annotations.

In conclusion I really don't want to do this so I hope someone else will :).

**Comment by ekrich** [comment]
Thank-you. With all that research it would seem that `coderbot` could do it.

**Comment by ekrich** [comment]
Wow, very cool - thanks.
